src/js/view/floating/floating-controller.ts

```diff
@@ -0,0 +1,303 @@
+import type Model from 'controller/model';
+import type Preview from 'view/preview';
+import type View from 'view/view';
+import type { FloatConfig } from 'controller/model';
+import FloatingDragUI from 'view/floating/floating-drag-ui';
+import { OS } from 'environment/environment';
+import { deviceIsLandscape } from 'utils/dom';
+import { isIframe } from 'utils/browser';
+import {
+    addClass,
+    removeClass
+} from 'utils/dom';
+import { STATE_IDLE, STATE_COMPLETE, STATE_ERROR } from 'events/events';
+import { isNumber } from 'utils/underscore';
+import {
+    style,
+} from 'utils/css';
+import viewsManager from 'view/utils/views-manager';
+
+const FLOATING_TOP_OFFSET = 62;
+let _floatingPlayer: HTMLElement | null = null;
+
+export default class FloatingController {
+    _playerEl: HTMLElement;
+    _parent: View;
+    _wrapperEl: HTMLElement;
+    _preview: Preview;
+    _model: Model;
+    _floatingUI: FloatingDragUI;
+    _mobileCheckCanFire?: boolean;
+    _mobileDebounceTimeout?: number;
+    _floatingStoppedForever: boolean;
+    _lastIntRatio: number;
+    _canFloat?: boolean;
+    boundThrottledMobileFloatScrollHandler: Function;
+
+    constructor(model: Model, parent: View, elements: { player: HTMLElement; wrapper: HTMLElement; preview: Preview }) {
+        this._playerEl = elements.player;
+        this._wrapperEl = elements.wrapper;
+        this._preview = elements.preview;
+        this._model = model;
+        this._parent = parent;
+        this._floatingUI = new FloatingDragUI(this._wrapperEl);
+        this._floatingStoppedForever = false;
+        this._lastIntRatio = 0;
+
+        this.boundThrottledMobileFloatScrollHandler = this.throttledMobileFloatScrollHandler.bind(this);
+    }
+    initFloatingBehavior(): void {
+        // Don't reinitialize this behavior if the user dismissed the floating player
+        if (this._floatingStoppedForever) {
+            return;
+        }
+        // Setup floating scroll handler
+        viewsManager.removeScrollHandler(this.boundThrottledMobileFloatScrollHandler);
+        if (this.getFloatingConfig()) {
+            const fm = this.getFloatMode();
+            if (fm === 'notVisible') {
+                if (OS.mobile) {
```

Comment:
not in love with this much nesting, maybe turn the parent `(this.getFloatingConfig())` conditional into an early return instead?